aten/src/ATen/native/sparse/SparseMatMul.cpp

```diff
@@ -0,0 +1,231 @@
+  result_indices.resize_({2, nnz_result});
+  result_values.resize_(values_size);
+
+  auto result_indices_accessor = result_indices.accessor<int64_t, 2>();
+  auto result_values_accessor = result_values.accessor<scalar_t, 1>();
+
+  int64_t index = 0;
+  for (auto kv : d) {
+    auto idx = kv.first;
+    result_indices_accessor[0][index] = idx.first;
+    result_indices_accessor[1][index] = idx.second;
+    result_values[index] = kv.second;
+    index++;
+  } 
+}
+
+template <typename scalar_t, bool grad_by_row>
+void sparse_matmul_kernel_grad(Tensor& result, const Tensor& x) {
+  auto x_indices = x._indices().contiguous();
+  auto x_values = x._values().contiguous();
+
+  auto nnz_a = x_values.size(0);
+
+  auto indices_accessor = x_indices.accessor<int64_t, 2>();
+  auto rows = indices_accessor[1];
+  auto cols = indices_accessor[0];
+  auto values_accessor = x_values.accessor<scalar_t, 1>();
+
+  auto result_indices = result._indices().contiguous();
+  auto result_values = result._values().contiguous();
+
+  auto n = result.size(0);
+  auto m = result.size(1);
+  auto size = grad_by_row ? n : m;
+  auto inner_size = grad_by_row ? m : n;
+
+  auto indices = (grad_by_row ? rows : cols);
+
+  std::vector<scalar_t> scalar_values(size, static_cast<scalar_t>(0));
+  for (int64_t i = 0; i < nnz_a; i++) {
+    for (int64_t index = 0; index < size; index++) {
+      if (indices[i] == index) {
+        scalar_values[index] += values_accessor[i];
+      }
+    }
+  }
+
+  int64_t index = 0;
+  std::map<std::pair<int64_t, int64_t>, scalar_t> d;
+
+  for (int64_t curr_index = 0; curr_index < scalar_values.size();
+       curr_index++) {
+    if (scalar_values[curr_index] != static_cast<scalar_t>(0)) {
+      for (int64_t other_index = 0; other_index < inner_size; other_index++) {
+        std::pair<int64_t, int64_t> current_index;
+        if (grad_by_row) {
+          current_index = std::make_pair(curr_index, other_index);
+        } else {
+          current_index = std::make_pair(other_index, curr_index);
+        }
+        d[current_index] += scalar_values[curr_index];
+        index++;
+      }
+    }
+  }
+  int64_t nnz_result = d.size();
+
+  std::vector<int64_t> values_size = {nnz_result};
+  result_indices.resize_({2, nnz_result});
+  result_values.resize_(values_size);
+
+  auto result_indices_accessor = result_indices.accessor<int64_t, 2>();
+  auto result_values_accessor = result_values.accessor<scalar_t, 1>();
+
+  index = 0;
+  for (auto kv : d) {
+    auto idx = kv.first;
+    result_indices_accessor[0][index] = idx.first;
+    result_indices_accessor[1][index] = idx.second;
+    result_values[index] = kv.second;
+    index++;
+  } 
+}
+
+} // end anonymous namespace
+
+Tensor sparse_matmul(const Tensor& self_, const Tensor& other_) {
+  TORCH_INTERNAL_ASSERT(self_.is_sparse());
+  TORCH_CHECK(self_.dim() == 2);
+  TORCH_CHECK(other_.dim() == 2);
+  TORCH_CHECK(self_.size(1) == other_.size(0), "Incompatible matrices");
```

Comment:
Nit: can this error message detail that the incompatibility is a dimensional mismatch and print the dimensions?